src/lib/db-listeners/index.ts

```diff
@@ -235,57 +235,51 @@ export class DBListeners extends BaseListener {
     }
   };
 
+  private throttledSaveDocument = throttle((document: DocumentModelType) => {
+    const { user } = this.db.stores;
+    const { key, changeCount, content } = document;
+    if (content) {
+      const updatePath = this.db.firebase.getUserDocumentPath(user, key, document.uid);
+      this.db.firebase.ref(updatePath)
+        .update({ changeCount, content: JSON.stringify(getSnapshot(content)) })
+        .then(() => {
+          // console.log("Successful save", "document:", key, "changeCount:", changeCount);
+        })
+        .catch(() => {
+          user.setIsFirebaseConnected(false);
+          console.warn("Failed save!", "document:", key, "changeCount:", changeCount);
+        });
+    }
+    // two-second throttle on the trailing edge so that with continued changes we save
+    // at least and at most once every two seconds
+  }, 2000, { trailing: true });
+
   private monitorDocumentModel = (document: DocumentModelType, monitor: Monitor) => {
     // skip if not monitoring local changes
     if (monitor !== Monitor.Local) {
       return;
     }
 
-    const { user } = this.db.stores;
     const { key, content } = document;
 
     const docListener = this.db.listeners.getOrCreateModelListener(`document:${key}`);
     if (docListener.modelDisposer) {
+      console.warn("Warning: monitorDocumentModel is monitoring a document that was already being monitored!",
+                    "key:", key, "contentId:", content?.contentId);
```

Comment:
In my investigation this became one of the most important indicators of problems. If things are working as expected we should never get here.